apis/v1beta1/gateway_types.go

```diff
@@ -0,0 +1,806 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=gateway-api,shortName=gtw
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="Class",type=string,JSONPath=`.spec.gatewayClassName`
+// +kubebuilder:printcolumn:name="Address",type=string,JSONPath=`.status.addresses[*].value`
+// +kubebuilder:printcolumn:name="Ready",type=string,JSONPath=`.status.conditions[?(@.type=="Ready")].status`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// Gateway represents an instance of a service-traffic handling infrastructure
+// by binding Listeners to a set of IP addresses.
+type Gateway struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of Gateway.
+	Spec GatewaySpec `json:"spec"`
+
+	// Status defines the current state of Gateway.
+	//
+	// +kubebuilder:default={conditions: {{type: "Scheduled", status: "Unknown", reason:"NotReconciled", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}}
+	Status GatewayStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// GatewayList contains a list of Gateways.
+type GatewayList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Gateway `json:"items"`
+}
+
+// GatewaySpec defines the desired state of Gateway.
+//
+// Not all possible combinations of options specified in the Spec are
+// valid. Some invalid configurations can be caught synchronously via a
+// webhook, but there are many cases that will require asynchronous
+// signaling via the GatewayStatus block.
+type GatewaySpec struct {
+	// GatewayClassName used for this Gateway. This is the name of a
+	// GatewayClass resource.
+	GatewayClassName ObjectName `json:"gatewayClassName"`
```

Comment:
Is there a reason to not just name the field `ClassName`? Or even just `GatewayClass`? The latter would be consistent with [`Service.Spec.LoadBalancerClass`](https://github.com/kubernetes/api/blob/b754a94214be15ffc8d648f9fe6481857f1fc2fe/core/v1/types.go#L4509-L4521). That said, the `Name` suffix _is_ used by [`Ingress.Spec.IngressClassName`](https://github.com/kubernetes/api/blob/cd271e7523ef7b6b485aa42c0218c294ca00a001/networking/v1/types.go#L258-L269).  In fact, the `Name` suffix is _also_ used by [`PersistentVolumeClaim.Spec.StorageClassName`](https://github.com/kubernetes/api/blob/b754a94214be15ffc8d648f9fe6481857f1fc2fe/core/v1/types.go#L348-L351), so it appears as if the outlier is `Service.Spec.LoadBalancerClass` and the convention _is_ to use the `Name` suffix. Carry on 😃 

Also, the comment could be simplified:

```go
// GatewayClassName is the name of the GatewayClass resource used by
// this Gateway.
```

Finally, the use of `ObjectName` is interesting. I see it has a validation on it for min and max length. However, [the comments](https://github.com/mehabhalodiya/gateway-api/blob/205398acbffdebcaacd89ebbd159b5356ee49576/apis/v1beta1/shared_types.go#L338-L344) on the type _also_ indicate additional, possible validations:

```go
// ObjectName refers to the name of a Kubernetes object.
// Object names can have a variety of forms, including RFC1123 subdomains,
// RFC 1123 labels, or RFC 1035 labels.
//
// +kubebuilder:validation:MinLength=1
// +kubebuilder:validation:MaxLength=253
```

I wonder if there is a reason to not just use `string` and let the webhook that will likely do the additional validation also be responsible for the min/max length validation?

Reply:
Hey @akutz! We're coming to the end of the v1beta1 release window, and both @robscott and I realized that your feedback had slipped through the cracks. So, I'll respond now where I think we probably won't make changes, and make issues for ones where we will.

For the min-max length validations and others, the general guideline we're using is to only add custom validation when we can't get it from the kubebuilder/OpenAPI validation.